### ISSUE TITLE
availability of surface variables?

### ISSUE THREAD

**Issue by GanZhang-GFD** [google-research/neuralgcm#108]
The NeuralGCM paper contains some analytics that depend on the surface information. For example, the precipitable water calculation involves a vertical integral that needs surface pressure. The tracking of cyclones uses sea-level pressure and near-surface wind. However, the output of the PressureLevelModel does not contain surface variables. Additionally, surface temperature is also a variable that can enable many applications. How may users access these surface variables?

**Comment by yaniyuval** [comment]
You can output the surface pressure from the model by adding to the config SurfacePressureDiagnostics.
Something like this should do the trick:
new_model_config_str = '\n'.join([
        ckpt['model_config_str'],
        'StochasticPhysicsParameterizationStep.diagnostics_module = @SurfacePressureDiagnostics',])
ckpt['model_config_str'] = new_model_config_str

Regarding TC tracking: we give a full detail in the appendix of our paper how we used 2 different methods (using sea level pressure/vorticity at 850hPa + other criterion) to track cyclones (that yielded similar results) but non-used near-surface wind.

We currently do not have the concept of "surface variables" (except surface pressure which is a prognostic variable in our dycore) and we output variables only on pressure levels.

**Comment by ShihengDuan** [comment]
Surface temperature would be useful.

**Comment by GanZhang-GFD** [comment]
Thanks @yaniyuval! Sorry for not seeing your reply earlier:

> You can output the surface pressure from the model by adding to the config SurfacePressureDiagnostics. Something like this should do the trick: new_model_config_str = '\n'.join([ ckpt['model_config_str'], 'StochasticPhysicsParameterizationStep.diagnostics_module = @SurfacePressureDiagnostics',]) ckpt['model_config_str'] = new_model_config_str

I placed the suggested edits between the step of downloading ckpt and "model = neuralgcm.PressureLevelModel.from_checkpoint(ckpt)". But the model initialization ran into the following error:

`ValueError: No configurable matching reference '@SurfacePressureDiagnostics'.`

Did I miss any other steps when configuring the model?
 
> Regarding TC tracking: we give a full detail in the appendix of our paper how we used 2 different methods (using sea level pressure/vorticity at 850hPa + other criterion) to track cyclones (that yielded similar results) but non-used near-surface wind.

SLP is the key variable. Thanks for the clarification about the tracking.

**Comment by GanZhang-GFD** [comment]
> Surface temperature would be useful.

Agree. Before the functionality is implemented, it is relatively straightfoward to approximate the surface temperature when surface pressure is available.

**Comment by yaniyuval** [comment]
@GanZhang-GFD ,
Try to pip install directly from the github page since the SurfacePressureDiagnostics was added only recently:
```pip install git+https://github.com/google-research/neuralgcm.git@main```

**Comment by GanZhang-GFD** [comment]
@yaniyuval Thanks! I managed to finish the run with the newer model version.

1) The surface pressure appears to be in the final state object. The data array does not have the unit attribute. The values suggest some log operation was involved. Reversing log10 yields a reasonable pattern, but the lowest values are too small. Did I miss any steps?

```
plt.contourf(10**(final_state.diagnostics['surface_pressure'].T), cmap='Reds')
plt.colorbar()
```
<img width="525" alt="Screenshot 2024-08-29 at 5 14 22 AM" src="https://github.com/user-attachments/assets/086d2ccf-9eac-4c25-81f2-0b20b5088724">

2) Is there a smart way to access surface pressure at steps before the final state?

**Comment by GanZhang-GFD** [comment]
@yaniyuval Thanks again! The code snippet didn't change the output values. 

The test was with the reference code with neuralgcm.git@main. I also tested other variants of the pressure unit with the model config dictionary. 

I am still interested in checking whether any existing mechanisms help access surface pressure at steps before the final state? Rolling out the model step by step is feasible but somewhat cumbersome.

**Comment by GanZhang-GFD** [comment]
Re drift in surface pressure: I also noticed drifts in other variables (e.g., temperature) of various magnitudes. Larger drifts usually come together with model instability. Smaller drifts appear to do relatively little harm when the simulations are not very long. I look forward to testing the next model version!

**Comment by yaniyuval** [comment]
@GanZhang-GFD ,
I slightly updated the snippet.
When I use it I get the correct units. See a snapshot. 
<img width="647" alt="Screenshot 2024-08-29 at 8 13 00 PM" src="https://github.com/user-attachments/assets/95426987-695f-453f-81e3-fda32059f14a">

**Comment by GanZhang-GFD** [comment]
@yaniyuval The updated snippet worked! Thanks.

For the others watching this issue: The snippet helps put the surface pressure into the prediction object (along with other variables like u, v, etc.). Since the surface pressure is avialable at the same output steps, my other question was also resolved.

**Comment by ShihengDuan** [comment]
> @GanZhang-GFD , Try to pip install directly from the github page since the SurfacePressureDiagnostics was added only recently: `pip install git+https://github.com/google-research/neuralgcm.git@main`

I'm still getting an error message ```No configurable matching reference '@SurfacePressureDiagnostics'.``` I've tried to pip install from the GitHub address.

**Comment by ankurmahesh** [comment]
@GanZhang-GFD , what method would you use to calculate surface temperature, given surface pressure, and the model state?  Thanks very much for any advice

**Comment by GanZhang-GFD** [comment]
> @GanZhang-GFD , what method would you use to calculate surface temperature, given surface pressure, and the model state? Would it be based on approximating a lapse rate? Thanks very much for any advice

@ankurmahesh With the surface pressure data, you may try to extrapolate the temperature on pressure levels to the surface. Your mileage may vary depending on your extrapolation technique and region of interest. For example, it can work reasonably with flat surfaces near the sea level but will struggle with complex terrains. In such cases, post-processing bias corrections are likely necessary.
